src/browser-locale/index.js

```diff
@@ -0,0 +1,64 @@
+/* global TERRA_AGGREGATED_LOCALES */
+
+// The fallback locale that will be used if a supported locale is unable to be retrieved from the browser.
+const DEFAULT_LOCALE = 'en';
+
+/**
+ * Returns a boolean indicating if the locale is supported.
+ * @param {string} locale - The locale.
+ * @returns {bool} - A boolean indicating the locale is supported.
+ */
+const isSupported = (locale) => {
+  if (typeof TERRA_AGGREGATED_LOCALES === 'object' && Array.isArray(TERRA_AGGREGATED_LOCALES) && locale) {
+    return TERRA_AGGREGATED_LOCALES.indexOf(locale) > -1;
+  }
+
+  return false;
+};
+
+/**
+ * Filters the user's preferred languages from the browser and returns the first language found that is included in the supported locale list.
+ * @returns {string|null} - A supported locale. Null if not found.
+ */
+const filterLanguages = () => {
+  if (navigator.languages && navigator.languages.length > 0) {
+    for (let index = 0; index < navigator.languages.length; index += 1) {
+      const locale = navigator.languages[index];
+
+      if (isSupported(locale)) {
+        return locale;
+      }
+    }
+  }
+
+  return null;
+};
+
+/**
+ * Retrieves the preferred browser locale.
+ * @returns {string} - A supported browser locale. Falls back to en.
+ */
+const getBrowserLocale = () => {
+  const preferredLocale = filterLanguages();
+
+  if (preferredLocale) {
+    return preferredLocale;
+  }
+
+  if (isSupported(navigator.language)) {
```

Comment:
According to https://html.spec.whatwg.org/multipage/system-state.html#dom-navigator-language

`navigator.language` should be returning the most preferred language. Should we be checking this value before filtering the languages?

Reply:
I don't think so. Browsing through various articles it suggests that the navigator.language is not a reliable way to obtain the user preference. From the Mozilla documentation we want the user defined languages to be used as first priority:

https://developer.mozilla.org/en-US/docs/Web/API/NavigatorLanguage